### ISSUE TITLE
Is this code tested?

### ISSUE THREAD

**Issue by rajdeep-engineer** [DawnGuardian/PGA460-ESP32_Library#2]
I have tried to run this code using ESP32 and Ultrasonic 2 click from Mikroe.
But the code doesn't work, even basic temperature reading also doesn't seem to work.

**Comment by DawnGuardian** [comment]
Yes, the code has been tested.

Please upload the wiring details between the modules, the relevant parts of the sketch, and the outputs/errors.

Also please specify the module of ESP32 used.

**Comment by rajdeep-engineer** [comment]
I used the Mikroe Ultrasonic 2 Click board [https://www.mikroe.com/ultrasonic-2-click](url)
with an ESP32 (DOIT ESP32 DEVKIT 1).
In the Ultrasonic 2 Click, I replaced the default transducer with "PUI Audio UTR-1440K-TT-R".

The hardware connections are as follows:-
1. 3V3 of Ultrasonic 2 Click -> 3.3V of ESP32
2. GND of Ultrasonic 2 Click -> GND of ESP32
3. RX of Ultrasonic 2 Click -> TX2 of ESP32 (GPIO17)
4. TX of Ultrasonic 2 Click -> RX2 of ESP32 (GPIO16)

I ran the "GetDistance" example from your repository. This is what I get in the terminal.
![Screenshot 2024-10-22 102722](https://github.com/user-attachments/assets/d85d94d9-bcf2-4f6d-a300-f91b71ad4a58)

**Comment by DawnGuardian** [comment]
There is a mistake in the readme where I have not mentioned the proper UART port to use. This not not the issue here because the port used in your setup is the correct one.

I'll probably write a change later on where you can pass a UART port when initialising the PGA460 object later.

On to the main issue, I've just tested the same example with a `DOIT ESP32 DEVKIT 1` and a module that I've created using the ESP32-S3. The PGA460 module I used is the Texas Instruments PGA460PSM-EVM. Both setups worked fine for me.

The thing to note is that the PGA460PSM-EVM requires a 5V input to power, not 3.3V. Seeing that the module you are using is also transformer based, I would suggest that the MikroE module also requires a 5V input.

**Comment by DawnGuardian** [comment]
Alright, so I just went through the documentation for the MikroE module. There is a minor mistake in the silkscreen labeling for the pins (TST has a 16 above it when it should be 1), but otherwise the pin descriptions seem fine.

There is an EN pin connected to the TLV61046A that should be driven high to enable the board. Please check if you have tied that pin to the 3V3 pin.

**Comment by rajdeep-engineer** [comment]
I test with a PGA460PSM-EVM.

Here are the hardware connections:-
1. PWR of PGA460PSM-EVM -> Vin of ESP32 (5V)
2. GND of PGA460PSM-EVM -> GND of ESP32
3. Rx of PGA460PSM-EVM -> Tx of ESP32  (GPIO17)
4. Tx of PGA460PSM-EVM -> Rx of ESP32  (GPIO16)

I am still not getting anything. Can you confirm if the hardware connection mappings are ok?

**Comment by DawnGuardian** [comment]
Yes, those hardware connections are correct. I do have additional pull ups for the UART lines (1k5 resistors).

If you are still only getting the output that had been posted previously, then the ESP32 has not connected to the PGA460 at all. My library does not currently check for a positive connection.

This lack of connection is usually because the PGA460 has not been powered properly, or a fault in the UART lines.

### Normal output
![Working output](https://github.com/user-attachments/assets/8bc74d8d-de82-4672-8f2b-0919cbf877dc)

Notice the timestamps show immediate return of data after the separation line.

### Output with PGA460 power removed
![Output with PGA460 power removed](https://github.com/user-attachments/assets/cf5217f4-e9e7-4e3a-aa4b-12fcded2f4fc)

Notice the timestamps now show a ~2 second delay after the separation line - this is due to the UART timing out.

**Comment by rajdeep-engineer** [comment]
ok thanks for the screenshots. 

I have checked the power pins of PGA460 and it seems fine. So, I am clueless why it's not working on both PGA460PSM-EVM and Mikroe Ultrasonic Click 2.

I am purchasing a BOOSTXL-PGA460 evaluation kit to study this further.

I had earlier written code for STM32 microcontroller to interface with PGA460. 
The strange thing I found is that PGA460 generated checksum is wrong (the checksum returned by PGA460 for against commands from MCU/controller). So, there is no way to tell if the data returned by PGA460 is correct or not.

The Energia code written by "Whitehead" also doesn't validate the checksum returned by PGA460.

**Comment by DawnGuardian** [comment]
Good luck, and please do get back if you figure out the issue. I'll also try to improve the library in the mean time to allow configuration of serial ports and connection detection.

And yes, the main motivation I had behind writing this library was how obfuscated the Energia library was. You'll see that I've actually separated the checksum process - not too great for speed, but then again, this is an Arduino library.

**Comment by rajdeep-engineer** [comment]
Sure, I will keep you posted.

**Comment by sfbence** [comment]
Hi, 
did you find out the cause of the issue? I am using the setup:
- BOOSTXL-PGA460 EVM, powered up by an external 12V power supply, power supply selector jumpers in "Standard" setup, (motherboard USB if connected, BOOSTXL from external 12V), power control LED is ON
- board detached from the "motherboard" from TI
- RX, TX from the ESP32-E (Firebeetle 2) on pins 16, 17
- the pins COM_SEL, COM_PD, MEM_HOLD, MEM_CS on other GPIOs, GNDs connected

The result is the same  screen with the -42.6667 C....
The same setup plugged on the motherboard and used with the GUI of the evaluation board works perfectly.
Could you give some further debugging hints?

I have made my custom transducer setup source code, and checked that although I don't have any error symptoms, I believe nothing is really written to the PGA460 UART, as after re-reading a register value I didn't get the same back.

Any hints how to doublecheck if Serial2 works? 

Thanks

**Comment by DawnGuardian** [comment]
@rajdeep-engineer Apologies, there might be a mistake with the hardware setup that I did not catch.

The PGA-460 (and Q1) chips operate on an input voltage of 6V to 18V - most ESP32 boards cannot provide this. The UART interface is still 3.3V - 5V.

# [PGA460PSM-EVM](https://www.ti.com/tool/PGA460PSM-EVM)
The VPWR pin must be connected to a 6V - 18V supply, it will not work on 5V.

# [BOOSTXL-PGA460](https://www.ti.com/tool/BOOSTXL-PGA460)
If using an external controller, **disconnect the MSP430 board**. The MSP430 has a higher driving power and will dominate the UART lines from your external controller.

Use the 'External Only' power setup described in page 8 of [this document](https://www.ti.com/lit/ug/slau659b/slau659b.pdf). From your external controller connect only the communications lines (UART in this case) and GND (to equalise signal reference).

Connect voltage supply only if you wish to power your external controller using the BOOSTXL board as your source.

The update with UART error reporting will deploy within the next week or so. Apologies again for the delay, they are only small changes but it is taking me some time due my workload. For now (and using this library), a report of zero values and temperature reading of -42.6667 C either means a UART connection fault or the PGA460 not being powered up.

**Comment by sfbence** [comment]
With further tests today I could close out some sources of errors: 
- when powered up with the external power supply, I could read the 5V on its 5V pin.
- my ESP32 can communicate fine through UART2  with other devices
- I was wondering if that could have an effect (but finally not): the PGA460 manual says, that on power-up the TEST pin decides, if it will use 3.3V or 5V for UART. I grounded the TEST to make sure it stays LOW to go for the 3.3V logic. 

Conclusion for the moment: board is powered up, but UART between ESP32 and BOOSTXL-PGA460 did not work. Later this week I will check the same, but with a PGA460 directly connected to the ESP32, that way the wiring issues and the BOOSTXL influence could be sorted out.

@DawnGuardian if you have a working setup, please share the exact wiring, setup of the BOOSTXL & code you are using

UPDATE: 
as I was giving up hope, and I have made a desperate last trial for finding out why it does not work, and reversed my UART wiring (using the BOOSTXL-PGA460 board), to a UARX->ESP32-RX and UATX->ESP32-TX wiring, and to my greatest surprise, it worked! Although, it still does not run on 115200 baud rate, but on a max 57600. If someone can explain why that is supposed to work, I would be very grateful.
Interestingly, when connecting a bare PGA460 chip, there it works again the expected TX->RX and RX->TX connection.

**Comment by DawnGuardian** [comment]
I have just updated this repo to make it a more general use library. It should be more responsive to use now, but I have also removed the specific pin controls needed for a TI dev board.

@sfbence :
I am so sorry - this was an issue I had discovered before and the correction I was supposed to have made to the readme was lost due to some circumstances.

Essentially, the jumpers show what the MSP430 dev board (that is supposed to be attached to the BOOSTXL) is configured to. So UARX refers to RX of the MSP430's RX and UATX refers to the TX of the MSP430's TX - they are not the RX and TX of the PGA460.

I am unsure why the baud seems limited to 57.6k, I have tested it at 115.2k baud and it works fine. If you have an MSP430 dev board connected, see if disconnecting it helps.
